### PR TITLE
Implement PASCAL'S TRIANGLE exercise.

### PR DESCRIPTION
**Pull request by daantjie**
Standard stuff.

### CODE REVIEW

**Comment by daantjie** [comment]
Hmm. Looks like I've buggered something up with the tests. I'll fix this as soon as I can.

File: exercises/pascals-triangle/example.lisp [hunk 1]
```diff
@@ -0,0 +1,16 @@
+(defpackage #:pascal
+  (:use #:common-lisp)
+  (:export #:rows))
+
+(in-package #:pascal)
+
+(defun row (n)
+  (if (eq n 1)
+      (list 1) ;prevents destructive functions from mutating the constant
+      (let ((prev (row (1- n))))
```
Comment (wobh) [inline exercises/pascals-triangle/example.lisp]: Can you reimplement this nonrecursively? Ideally with `loop' and a little algebra. Thanks!

File: exercises/pascals-triangle/pascal-test.lisp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+(define-test three-rows
+    (assert-equal (pascal:rows 3)
+                  '((1) (1 1) (1 2 1))))
+
+(define-test four-rows
+    (assert-equal (pascal:rows 4)
+                  '((1) (1 1) (1 2 1) (1 3 3 1))))
+
+(define-test negative-rows
```
Comment (wobh) [inline exercises/pascals-triangle/pascal-test.lisp]: I suggest leaving this case undefined. Pascal's triangle can be extended into negative numbers, but I I think that's best left optional.

File: exercises/pascals-triangle/example.lisp [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  (loop for i from 0 to n
+        for fact = 1 then (* fact i)
+        finally (return fact)))
+
+(defun choose (n r)
+  (/ (fact n)
+     (* (fact r) (fact (- n r)))))
+
+(defun rows (n)
+  (if (> n 0)
```
Comment (wobh) [inline exercises/pascals-triangle/example.lisp]: If this were an excercism submission, I'd suggest using `when` instead of `if` here.

File: exercises/pascals-triangle/pascal-test.lisp (hunk 2 above)
Comment (daantjie) [inline exercises/pascals-triangle/pascal-test.lisp]: The exercise spec gives negative rows – is it fine to ignore it? (I would personally agree with you.)

Comment (verdammelt) [inline exercises/pascals-triangle/pascal-test.lisp]: I'd say leave it. I'm a little out of the loop with the larger discussion of consistency in tests across tracks - but I think the desire is still there to keep them consistent.

Did you create these tests based upon the common test data?

**Comment by verdammelt** [comment]
For this new exercise submission we should also decide upon a difficulty level (or leave it 1 like all the others) and topics.

File: exercises/pascals-triangle/pascal-test.lisp (hunk 2 above)
Comment (daantjie) [inline exercises/pascals-triangle/pascal-test.lisp]: @verdammelt; yup, pretty much copy-and-paste with a little tweaking.

**Comment by verdammelt** [comment]
To keep this PR from lingering:

- [ ] Changing usage of `assert-equal` to be `expected actual` not `actual expected`
- [ ] Decide upon a exercise level (which could simply be leave it at `1`
- [ ] Decide upon exercise topic(s)

If we do those things I'd merge the PR.

Thoughts?

**Comment by wobh** [comment]
Fixing the tests will fix the build.

Topics and difficulty are already set in the latest commit.

That makes for a really short checklist!

**Comment by verdammelt** [comment]
I guess I needed to add:

- [ ] Mark: make sure you review the damn PR

**Comment by daantjie** [comment]
Okay, I've changed the order of args in `assert-equal` – why does that make a difference to the tests?

**Comment by verdammelt** [comment]
@daantjie The library `lisp-unit` specifies that it should be `expected actual` which probably means that when they produce error messages it will read that way.

@wobh was hoping it would cause the build to pass - but it still doesn't look like it does.

The build doesn't seem to be providing any details with the failure:

> ; loading #P"/home/travis/build/exercism/common-lisp/exercises/pascals-triangle/example.lisp"
; #<PACKAGE "PASCAL">
; #<PACKAGE "PASCAL">
; FACT
; CHOOSE
; ROWS
ALERT: Recording problem

So I am unsure what is going wrong at the moment. I'm assuming @daantjie that the build runs fine on your machine?

**Comment by daantjie** [comment]
Actually, they don't pass if I use the `ros` version.

**Comment by verdammelt** [comment]
Our build should be loading `example.lisp` before running the tests. 

Does your build failure with `ros` look the same as the one on Travis? Can you narrow down which test(s) are failing?

**Comment by wobh** [comment]
I'm investigating the testing issue and have reproduced it locally. Let me say, THANK YOU to @daantjie for your pull-request and patience. This has surfaced at least one bug in `xlisp-test`. PR on that coming soon.

**Comment by verdammelt** [comment]
Yes thank you @daantjie for finding a problem with our build, and also your patience. I want to integrate your PR!

Also thank you @wobh for jumping in to figure this out!

🍸  cheers to you both!

**Comment by wobh** [comment]
@verdammelt would you merge #159 when convenient

@daantjie When #159 is merged 

1. branch your changes for the Pascal's triangle exercise
2. `git pull upstream master` and rebase branch changes on top of it
3. After this you should be able to run `ros run -l 'bin/xlisp-test.lisp' -e '(xlisp-test:test-exercise "pascals-triangle" 7)' -q`
4. regardless of the output of step 3, add a test to `pascal-tests` that definitely fails, for example: `(define-test fail (assert-true nil))`

Comment here to let us know what happens when you run the local tests (they should fail).

**Comment by daantjie** [comment]
Okay! I'll watch for #159.

**Comment by verdammelt** [comment]
#159 merged and closed.

**Comment by daantjie** [comment]
After (3),
```
$ ros run -l 'bin/xlisp-test.lisp' -e '(xlisp-test:test-exercise "pascals-triangle" 7)' -q
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/example.lisp"
; #<PACKAGE "PASCAL">
; #<PACKAGE "PASCAL">
; FACT
; CHOOSE
; ROWS
DEBUG: Loaded package #<PACKAGE "PASCAL">
DEBUG: Loaded example #<PACKAGE "PASCAL">
ALERT: Recording problem
```
and after adding the null test, the same output:
```
$ ros run -l 'bin/xlisp-test.lisp' -e '(xlisp-test:test-exercise "pascals-triangle" 7)' -q
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/example.lisp"
; #<PACKAGE "PASCAL">
; #<PACKAGE "PASCAL">
; FACT
; CHOOSE
; ROWS
DEBUG: Loaded package #<PACKAGE "PASCAL">
DEBUG: Loaded example #<PACKAGE "PASCAL">
ALERT: Recording problem
```

**Comment by wobh** [comment]
@daantjie would you push your new branch?

**Comment by daantjie** [comment]
@wobh Done!

**Comment by wobh** [comment]
Okay, so the problem is definitely that the test file name should be the same as the slug name, plus "-test". Here' it's named "pascal-test", but it should be named "pascals-triangle-test". The tests file just isn't getting loaded because the test runner expects it to be named after the slug.

The reason why this doesn't cause a problem is that `xlisp-test:test-exercises` wraps the loading of tests in a `handler-case` to make sure errors get reported rather than halting the whole works. This reporting isn't happening properly, which is a tangle of it's own. I'll make a separate issue for it.

Next steps here is to change the file name and run your tests. You should see the `fail` test fail. When you see that delete the `fail` test and rerun. I recommend squashing your commits, but that's optional.

**Comment by wobh** [comment]
TIL http://chem-bla-ics.blogspot.com/2011/01/github-tip-download-commits-as-patches.html

**Comment by daantjie** [comment]
Doesn't seem to be working for me...
```
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/example.lisp"
; #<PACKAGE "PASCALS-TRIANGLE">
; #<PACKAGE "PASCALS-TRIANGLE">
; FACT
; CHOOSE
; ROWS
DEBUG: Loaded package #<PACKAGE "PASCALS-TRIANGLE">
DEBUG: Loaded example #<PACKAGE "PASCALS-TRIANGLE">
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/pascals-triangle-test.lisp"
To load "lisp-unit":
  Load 1 ASDF system:
    lisp-unit
; Loading "lisp-unit"

; ("lisp-unit")
; #<PACKAGE "PASCALS-TRIANGLE-TEST">
; #<PACKAGE "PASCALS-TRIANGLE-TEST">
; FAIL
ALERT: Recording problem
```

**Comment by wobh** [comment]
It loaded the test file which it wasn't doing before.

**Comment by daantjie** [comment]
@wobh True! Do you have any idea what xlisp means by the ‘recording problem’ (or do you think the error's caused by something else)?

**Comment by wobh** [comment]
@daantjie It's a bit of logging code that prints when an error occurs and the error is being recorded.

If you change the line 106 to  `(alert "Recording problem ~S" result)` it may print out something more useful.

However, I'm not sure. I took a patch of this PR, applied it to master, renamed the test file and the tests failed with the single failure I expected.

**Comment by daantjie** [comment]
It's possible that I've not renamed them properly. How about creating a new branch off the patch for my commits, then applying the following patch (which is what I did).

https://gist.github.com/daantjie/61d83fa7d1f88c06a1156500603821af

Or you could use curl:
```
$ curl https://gist.githubusercontent.com/daantjie/61d83fa7d1f88c06a1156500603821af/raw/223a9abf75c80dc7bb5551b157bdf62f2544fd9f/rename-tests.patch > rename-tests.patch
```

**Comment by wobh** [comment]
I'll look at that gist more closely to try to understand what's going wrong there. However, this is all I had to do to make it work (well, fail, where expected). 

https://github.com/wobh/xlisp/commit/e5dbd87509c215c62095bee532fef329b6f3d0c3

Expected results for SBCL:

https://travis-ci.org/wobh/xlisp/jobs/251305835#L1604

When you're seeing this, you can delete the fail-sure test, commit and push.

**Comment by daantjie** [comment]
Okay, stashing my botched changes and applying yours,
```
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/example.lisp"
; #<PACKAGE "PASCAL">
; #<PACKAGE "PASCAL">
; FACT
; CHOOSE
; ROWS
DEBUG: Loaded package #<PACKAGE "PASCAL">
DEBUG: Loaded example #<PACKAGE "PASCAL">
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/pascals-triangle-test.lisp"
To load "lisp-unit":
  Load 1 ASDF system:
    lisp-unit
; Loading "lisp-unit"

; ("lisp-unit")
; #<PACKAGE "PASCAL-TEST">
; #<PACKAGE "PASCAL-TEST">
; FAIL
; ZERO-ROWS
; SINGLE-ROW
; TWO-ROWS
; THREE-ROWS
; FOUR-ROWS
; NEGATIVE-ROWS
DEBUG: Loaded package #<PACKAGE "PASCAL-TEST">
DEBUG: Loaded exercise tests #<PACKAGE "PASCAL-TEST">
INFO: Running tests for #<PACKAGE "PASCAL">
 | Failed Form: NIL
 | Expected T but saw NIL
 |
FAIL: 0 assertions passed, 1 failed.

ALERT: Recording problem
```

and, commenting out the `fail` test,

```
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/example.lisp"
; #<PACKAGE "PASCAL">
; #<PACKAGE "PASCAL">
; FACT
; CHOOSE
; ROWS
DEBUG: Loaded package #<PACKAGE "PASCAL">
DEBUG: Loaded example #<PACKAGE "PASCAL">
; loading #P"/home/daniel/code/exercism/common-lisp/exercises/pascals-triangle/pascals-triangle-test.lisp"
To load "lisp-unit":
  Load 1 ASDF system:
    lisp-unit
; Loading "lisp-unit"

; ("lisp-unit")
; #<PACKAGE "PASCAL-TEST">
; #<PACKAGE "PASCAL-TEST">
; ZERO-ROWS
; SINGLE-ROW
; TWO-ROWS
; THREE-ROWS
; FOUR-ROWS
; NEGATIVE-ROWS
DEBUG: Loaded package #<PACKAGE "PASCAL-TEST">
DEBUG: Loaded exercise tests #<PACKAGE "PASCAL-TEST">
INFO: Running tests for #<PACKAGE "PASCAL">
```

**Comment by wobh** [comment]
That looks good!

**Comment by daantjie** [comment]
- [x] squashed commits
- [x] pushed changes

@wobh @verdammelt Thank you so much for patiently guiding me through this process! I hope to implement some more exercises soon (and have them pass first time!).

File: exercises/pascals-triangle/pascals-triangle-test.lisp [hunk 4]
```diff
@@ -0,0 +1,37 @@
+(ql:quickload "lisp-unit")
+#-xlisp-test (load "pascal")
+
+(defpackage #:pascal-test
+  (:use #:common-lisp #:lisp-unit))
+(in-package #:pascal-test)
+
+;; (define-test fail
```
Comment (wobh) [inline exercises/pascals-triangle/pascals-triangle-test.lisp]: Just delete this.

**Comment by daantjie** [comment]
I think I've mucked the commits up a bit – tell me if it's a big problem, and I'll try to figure out how to fix it.

**Comment by wobh** [comment]
You should be able to rebase, deleting the `a513cd9` commit and marking `2e5b322` as "fixup" or "f" to squash it into the base commit and throwing away the commit message.

**Comment by verdammelt** [comment]
@daantjie thank *you* for your patience, and for submitting a new exercise... and help with the QA of the build system :)

**Comment by wobh** [comment]
I didn't see when this got cleaned up. Sorry. Merging. Thank you @daantjie!

**Comment by daantjie** [comment]
@wobh; Cheers!
